fmpz/multi_crt.c

```diff
@@ -11,346 +11,338 @@
     }
 
-    /* two more spots for temporaries */
-    P->temp1loc = P->localsize++;
-    P->temp2loc = P->localsize++;
-
-    TMP_END;
-
-    return P->good;
-}
-
-int fmpz_multi_crt_precompute(
-    fmpz_multi_crt_t P,
-    const fmpz * moduli,
-    slong len)
-{
-    int success;
-    slong i;
-    const fmpz ** m;
-    TMP_INIT;
+    for (i = r, j = 0; j < 2*r - 4; i++, j += 2)
+    {
+        slong s, minp;
+        const fmpz * mind;
 
-    FLINT_ASSERT(len > 0);
+        minp = j;
+        mind = v + j;
+        for (s = j + 1; s < i; s++)
+        {
+            if (fmpz_cmp(v + s, mind) < 0)
+            {
+                mind = v + s;
+                minp = s;
+            }
+        }
+        fmpz_swap(v + j, v + minp);
+        SLONG_SWAP(link[j], link[minp]);
 
-    TMP_START;
+        minp = j + 1;
+        mind = v + j + 1;
+        for (s = j + 2; s < i; s++)
+        {
+            if (fmpz_cmp(v + s, mind) < 0)
+            {
+                mind = v + s;
+                minp = s;
+            }
+        }
+        fmpz_swap(v + j + 1, v + minp);
+        SLONG_SWAP(link[j + 1], link[minp]);
 
-    m = (const fmpz **) TMP_ALLOC(len*sizeof(fmpz *));
-    for (i = 0; i < len; i++)
-    {
-        m[i] = moduli + i;
+        fmpz_mul(v + i, v + j, v + j + 1);
+        link[i] = j;
     }
 
-    success = fmpz_multi_crt_precompute_p(P, (const fmpz * const *) m, len);
+    fmpz_mul(P->final_modulus, v + 2*r - 4, v + 2*r - 3);
 
-    TMP_END;
+    fmpz_one(one);
+    P->good = _fill_pfrac(link, v, w, 2*r - 4, one, g, s, t);
+    if (P->good)
+        _fill_prog(P, link, v, w, 2*r - 4, 0);
 
-    return success;
-}
+    fmpz_clear(one);
+    fmpz_clear(g);
+    fmpz_clear(s);
+    fmpz_clear(t);
 
+    for (i = 0; i < 2*(2*r - 2); i++)
+        fmpz_clear(v + i);
 
+    flint_free(link);
+    flint_free(v);
 
-void fmpz_multi_crt_precomp(
-    fmpz_t output,
-    const fmpz_multi_crt_t P,
-    const fmpz * inputs)
-{
-    slong i;
-    fmpz * out;
-    TMP_INIT;
-
-    TMP_START;
-    out = (fmpz *) TMP_ALLOC(P->localsize*sizeof(fmpz));
-    for (i = 0; i < P->localsize; i++)
-    {
-        fmpz_init(out + i);
-    }
+done:
 
-    fmpz_swap(out + 0, output);
-    _fmpz_multi_crt_run(out, P, inputs);
-    fmpz_swap(out + 0, output);
+    P->temp1loc = P->localsize++;
+    P->temp2loc = P->localsize++;
+    P->temp3loc = P->localsize++;
+    P->temp4loc = P->localsize++;
 
-    for (i = 0; i < P->localsize; i++)
+    if (!P->good)
     {
-        fmpz_clear(out + i);
+        fmpz_one(P->final_modulus);
+        P->length = 0;
     }
 
-    TMP_END;
+    return P->good;
 }
 
-void fmpz_multi_crt_precomp_p(
+
+void fmpz_multi_CRT_precomp(
     fmpz_t output,
     const fmpz_multi_crt_t P,
-    const fmpz * const * inputs)
+    const fmpz * inputs,
+    int sign)
 {
     slong i;
     fmpz * out;
     TMP_INIT;
 
     TMP_START;
-    out = (fmpz *) TMP_ALLOC(P->localsize*sizeof(fmpz));
+    out = TMP_ARRAY_ALLOC(P->localsize, fmpz);
```

Comment:
Is this TMP_ARRAY_ALLOC ensuring stack overflows are unlikely?

Reply:
The function is not recursive. If I can't use TMP_ALLOC here, then I don't know of any place where I can use it.

Reply:
Not quite what I meant, but ok.